src/module.c

```diff
@@ -1074,213 +1072,535 @@ int RM_CreateSubcommand(RedisModuleCommand *parent, const char *name, RedisModul
+ *     * `REDISMODULE_ARG_TYPE_INTEGER`: Integer argument.
+ *     * `REDISMODULE_ARG_TYPE_DOUBLE`: Double-precision float argument.
+ *     * `REDISMODULE_ARG_TYPE_KEY`: String argument representing a keyname.
+ *     * `REDISMODULE_ARG_TYPE_PATTERN`: String, but regex pattern.
+ *     * `REDISMODULE_ARG_TYPE_UNIX_TIME`: Integer, but Unix timestamp.
+ *     * `REDISMODULE_ARG_TYPE_PURE_TOKEN`: Argument doesn't have a placeholder.
+ *       It's just a binary token. (TODO: Add example.)
+ *     * `REDISMODULE_ARG_TYPE_ONEOF`: Used when user can choose only one of a
+ *       few sub-arguments. Requires `subargs`. (TODO: Add example.)
+ *     * `REDISMODULE_ARG_TYPE_BLOCK`: Used when one wants to group together
+ *       several sub-arguments, usually to apply something on all of them (like
+ *       making the entire group "optional"). Requires `subargs`. (TODO: Add
+ *       example.)
+ *
+ *     Explanation of the command argument flags:
+ *
+ *     * `REDISMODULE_CMD_ARG_OPTIONAL`: The argument is optional (like GET in
+ *       the SET command).
+ *     * `REDISMODULE_CMD_ARG_MULTIPLE`: The argument may repeat itself (like
+ *       key in DEL).
+ *     * `REDISMODULE_CMD_ARG_MULTIPLE_TOKEN`: The argument may repeat itself,
+ *       and so does its token (like `GET pattern` in SORT).
+ *
+ * On success REDISMODULE_OK is returned. On error REDISMODULE_ERR is returned
+ * and `errno` is set to EINVAL if invalid info was provided or EEXIST if info
+ * has already been set. */
+int RM_SetCommandInfo(RedisModuleCommand *command, const RedisModuleCommandInfo *info) {
+    if (!moduleValidateCommandInfo(info)) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    }
 
-/* Returns 1 if `cmd` is a command of the module `modulename`. 0 otherwise. */
-int moduleIsModuleCommand(void *module_handle, struct redisCommand *cmd) {
-    if (cmd->proc != RedisModuleCommandDispatcher)
-        return 0;
-    if (module_handle == NULL)
-        return 0;
-    RedisModuleCommand *cp = (void*)(unsigned long)cmd->getkeys_proc;
-    return (cp->module == module_handle);
-}
+    struct redisCommand *cmd = command->rediscmd;
 
-void extendKeySpecsIfNeeded(struct redisCommand *cmd) {
-    /* We extend even if key_specs_num == key_specs_max because
-     * this function is called prior to adding a new spec */
-    if (cmd->key_specs_num < cmd->key_specs_max)
-        return;
+    /* Check if any info has already been set. Overwriting info involves freeing
+     * the old info, which is not implemented. */
+    if (cmd->summary || cmd->complexity || cmd->since || cmd->history ||
+        cmd->hints || cmd->args ||
+        !(cmd->key_specs_num == 0 ||
+          /* Allow key spec populated from legacy (first,last,step) to exist. */
+          (cmd->key_specs_num == 1 && cmd->key_specs == cmd->key_specs_static &&
+           cmd->key_specs[0].begin_search_type == KSPEC_BS_INDEX &&
+           cmd->key_specs[0].find_keys_type == KSPEC_FK_RANGE))) {
```

Comment:
maybe in order to encourage people to stop using the legacy range spec we should fail this API if RM_CreateCommand was called with anything other than `0,0,0` for the range spec?

it will also simplify this code block ^

Reply:
@oranagra WDYT?

Reply:
If keyspecs are provided in the info struct, I think it makes sense to require legacy (0,0,0). Otherwise, info doesn't touch keyspecs and then I don't think we should fail.

Reply:
maybe i misunderstood that `if`

so what's you're saying is that we will fail if:
1. user used the legacy range in RM_CreateCommand (i.e. not 0,0,0); and
2. user gave key specs in RM_SetCommandInfo
is that true? if so, do we already have that check in the code?

other than that I guess you're right... if user used the legacy range and just wants to set the summary or something, we should not fail it

Reply:
i'm not sure i understand everything (short in time), but i'd like to raise this concern (which we did discuss in the past).

a new module is compiled with a new redismodule.h, but that module wants to still be ABI compatible with old redis versions.
so it registers its commands with the old API, and then calls the new one only if it's non-NULL.
this module will still wanna pass the legacy key-spec to the initial command registration.

Reply:
@oranagra That's what the PR does now. (The if statement here just checks that setcommandinfo hasn't been called before for the same command, mainly to avoid freeing the old allocated info.)

@guybe7 agree? If yes, resolve?

Reply:
ok, resolved (for some reason the "resolve" button disappeared.. i can only comment)